Clamp

silviacasac opened the pull request:
First pull request for the Clamp transformation. It has undergone a first correction by Prof. Vadhan on June 22 and a first correction by Mike on July 1 and 7. All these corrections have been implemented. The pseudocode has not been changed to include Andy's comment on July 1 on adding the domains and metrics as declarative statements in the preconditions.

raprasad reviewed (approved):
Looks good!

pdurbin commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,341 @@
+% OLD VERSION: The current OpenDP library contains the \texttt{make\_clamp\_vec} function implementing the clamping function. This is defined in lines 68-80 of the file \texttt{clamp.rs} (\url{https://github.com/opendp/opendp/blob/21-impute/rust/opendp/src/trans/clamp.rs}).
+
+\subsection{Code in Rust}
+The current OpenDP library contains the \texttt{make\_clamp\_vec} function implementing the clamping function. This is defined in lines 25-38 of the file \texttt{manipulation.rs} in the Git repository\footnote{As of June 16, 2021. Since then, the code has been updated to include a more general clampable domain, which is not yet finished.} (\url{https://github.com/opendp/opendp/blob/58feb788ec78ce ...
```
For now the commit hash is fine but in the future, once releases of OpenDP have been made, it might be nice for readability to link to lines 25-38 of version 1.0 or whatever.

silviacasac replied:
Good point, we will remember to change this

pdurbin commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,341 @@
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\mike{The input metric can just be hardcoded to SymmetricDistance. This means there's no generic MI, so the precondition can be dropped} \silvia{Same for output metric, no? Also, wouldn't it be safer not to hardcode it in case more metrics are added to the library in the future?}
+\mike{After recent talks, I think we're thinking this should go in the preconditions. We can expand the proof in the future if we get more metrics.} \silvia{Agreed}
+
+\mike{The type T can be any type that implements PartialOrd- for example floats or ints}
+
+\silvia{Fixed, but I think it should be ``implements TotalOrd", no?}
+\mike{Yes ^^. I'm still seeing references to float throughout this document though. You should be able to just refer to T, where T is qualified by the trait bounds.} \silvia{Agreed}
+
+\silvia{The domain/metric preconditions will be added once Prof. Vadhan agrees}
```
This conversation between Mike and Silvia (and various comments) should be removed before the proof is merged, right? Right now they are visible in the PDF.

Shoeboxam replied:
Agreed, I think we should generally delete these comments, once resolved, now that we have this review system.

silviacasac replied:
I have removed them in the version I just uploaded. There are two of my comments left because they are not fully closed topics (although very small)

pdurbin commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,341 @@
+\textbf{(Appropriate output domain).} In the case of \texttt{MakeClamp}, this corresponds to showing that for every vector $v$ of type \texttt{Vec(T)}, $\function(v)$ is an element of \texttt{VectorDomain} \texttt{(IntervalDomain(L, U))}. For that, we need to show two things: first, that the vector entries of $\function(v)$ have type \T. % Recall that the compiler infers the float type from Vecto ...
+Second, that they belong to the interval $[L, U]$.
+
+That $\function(v)$ has type \texttt{Vec(T)}  follows from the assumption that $v$ is in the input domain, the precondition requirement that \T has type \T, and the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(T)} and returns an element of type \texttt{Vec(T)}. If the Rust code compiles co ...
```
In the PDF, `\T has` looks a bit squished together as "Thas":

<img width="117" alt="Screen Shot 2021-07-08 at 12 05 01 PM" src="https://user-images.githubusercontent.com/21006/124955156-d5f3e300-dfe4-11eb-9e06-9a89b75c872c.png">

Later such squishing include "Tis" and "Tby".

silviacasac replied:
Fixed, thank you for the catch

pdurbin commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,341 @@
+% More proofs? Mike: Most proofs give conditions for which they are valid on. This mostly captures domains and metrics. You need a proof for the relation-- the function pseudocode, domains and metrics are all necessary to qualify the proof. I guess it may also make sense to have a proof for the output domain.
+
+% Not sure: separate into vec and sensitivity?
+
```
At this point in the PDF (page 6) we see "OLD VERSION BEFORE PROF. VADHAN’S CORRECTIONS – IGNORE". I assume before this is merged we want all of this gone from the PDF.

silviacasac replied:
Yes, I have now uploaded a PDF that contains only the relevant pages

pdurbin reviewed (commented):
Good stuff. The math is beyond be but I left a few comments.

silviacasac commented:
New version includes the following changes:
- Resolved comments by Phil and Mike above.
- Added the preconditions (and taken them out from the pseudocode) as discussed last week. We also decided to separate the preconditions into "user-specified types" and the 4 "input metric, output metric, input metric, output metric" that all transformations will have. We also decided to add a postcondition to deal with the fallible case.
- Changed floats to non-null floats after Mike's comment on total ordering not working for possibly null floats.
- Changed the statement of Theorem 1 to make it as general as possible (i.e., "under input_metric" instead of "under symmetric distance" so that the Theorem statement can be re-used in the proofs. Changed "appropriately quantified" to specifying the types of din, dout; which is what it meant.
- Changed some of the "rustiness" expressions that were left in the proof after the preconditions explained "in words" change (e.g., delete "VectorDomain(IntervalDomain(L, U))".
- Deleted the section "Conditions as specified in the pseudocode" because it is repetitive with the new preconditions. There is also no equation form for the stability relation and I refer directly to the pseudocode lines.
- Added table of contents

Further questions on the new version:
- Type sign L and U as it is now ("def MakeClamp(L: T, U: T)" or add that L and U have type T in the preconditions?
- Move the "if L>U: raise Exception" to the preconditions? a) But this would not be a compile time error. b) Mike said that the Rust construction for IntervalDomain is enforced to check that L<=U, so do I need to check this condition in the pseudocode and raise an exception if IntervalDomain(L, U) is in the preconditions?
- We were discussing to maybe define symmetric distance for vectors (as opposed to multisets) so that we can drop having to write MultiSet(v) every time in the proofs.

silviacasac commented:
New version (July 14) includes the following changes:
- Domains and metrics are still listed as preconditions (pending a double-check with @andrewvyrros), but they are likely to be put back into the code, as discussed with @SalilVadhan on July 13. Main issue: the declaration of certain domains requires variables that are only defined in the pseudocode.
- Added another part to the theorem statement (which should, from now on, appear in all of the proofs): "2. Domain-metric compatibility". As discussed on July 13, it is necessary to ensure that the input domain is part of the list of domains included in the definition of said metric. This usually follows directly from our list of definitions, but some later proofs might need to include some checking in the code.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+        \item Type \texttt{T} must have trait \texttt{TotalOrd}.\footnote{For now, the OpenDP library only implements \texttt{PartialOrd}, but \texttt{TotalOrd} will soon be implemented.}
+    \end{itemize}
+    \item \texttt{input\_domain:} any vector of elements of type \texttt{T}.
+    \item \texttt{output\_domain:} any vector of elements of type \texttt{T} which are contained in the interval \texttt{[L, U]}, where 
```
Given the dependence on L and U, it seems best to do this in the code, and then for consistency we probably also want to move the input_domain, input_metric, and output_metric into the pseudocode too.

silviacasac replied:
Agreed -- since I made the PR last week when we were still discussing about this I kept the domains and metrics in the preconditions for clamp (but already moved them to the code in the case of bounded sum). In the version I just updated they are now moved back into the pseudocode.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+
+\subsubsection*{Postconditions}
+\begin{itemize}
+    \item A \texttt{Transformation} is returned (i.e., if a \texttt{Transformation} cannot be returned successfully, then an error should be returned).
```
I'd say "Either a *valid* Transformation is returned or an error is returned".  The definition of what constitutes a valid transformation should appear in the pseudocode_defs file.

silviacasac replied:
Changed, thank you!

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Type \texttt{T} must have trait \texttt{TotalOrd}.\footnote{For now, the OpenDP library only implements \texttt{PartialOrd}, but \texttt{TotalOrd} will soon be implemented.}
```
add a cross-reference to the pseudocode_defs file so that the reviewer knows where to find the def of the trait TotalOrd.

silviacasac replied:
We have been meaning to add cross-references between the tex files, but it didn't quite work (tex-wise) -- we have to look again into it. But in any case, are you then saying that we should cross-reference every single term that appears defined in pseudocode_defs and proof_defs? So what I mean is that it is not just TotalOrd, but also many other terms in the document

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+The current OpenDP library contains the \texttt{make\_clamp\_vec} function implementing the clamping function. This is defined in lines 25-38 of the file \texttt{manipulation.rs} in the Git repository\footnote{As of June 16, 2021. Since then, the code has been updated to include a more general clampable domain, which is not yet finished.} (\url{https://github.com/opendp/opendp/blob/58feb788ec78ce ...
+
+\begin{figure}[ht]
+    \includegraphics[width=16cm]{clamp_rust.png}
```
is this an old version of the code?  still has PartialOrd in it.

SalilVadhan replied:
Also it looks like it doesn't enforce that M is SymmetricDistance.

silviacasac replied:
No, TotalOrd is still not implemented in the OpenDP code. Mike said it is in the to do list, but it has not yet been changed, which is why PartialOrd still appears in the pseudocode. Still, I do the proof with TotalOrd because Mike said it will for sure be updated in the near future.

SalilVadhan replied:
Got it, thanks

Shoeboxam replied:
I haven't made the change to Ord a priority because it hasn't been clear to me if this is what we want for the library. I wrote a bit more about this in an issue here: https://github.com/opendp/opendp/issues/203

I'd be happy to make the changes once it's clear that we want to adopt noisy-float newtypes to strengthen this trait bound.

silviacasac replied:
(Discussed on the architecture meeting on August 3rd)

SalilVadhan replied:
Until the code has been fixed, the proof document should say explicitly that the proof is being given assuming Ord even though the Rust code only requires PartialOrd, so the Rust code is not quite correct yet.   In general, any gaps like this should be highlighted clearly at the beginning of the document under a section titled something like "Remaining Gaps between Proof and Code".

silviacasac replied:
It is a good idea to have such a section -- but TotalOrd is fixed now, so I deleted it

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+    \end{itemize}
+    \item \texttt{input\_domain:} any vector of elements of type \texttt{T}.
+    \item \texttt{output\_domain:} any vector of elements of type \texttt{T} which are contained in the interval \texttt{[L, U]}, where 
+    \texttt{L} and \texttt{U} are of type \texttt{T}. \silvia{Problem: \texttt{L} and \texttt{U} appear only later in the pseudocode, as discussed on July 13.}
```
we agreed that these will move back into the pseudocode, right?

silviacasac replied:
Yes, sorry this is from last week. It is now updated accordingly.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+    
+    def function(data: Vec(T)) -> Vec(T): |\label{line:fn}|
+        def clamp(x: T) -> T: |\label{line:clamp}|
+            return max(min(x, U), L)
+        return list(map(clamp, data)) |\label{line:map}|
```
where can I find definitions of max, min, list, map?  Are these built-in Rust functions?

silviacasac replied:
Max and min are defined in the pseudocode definitions file (see Definitions 4.17 and 4.18). Map and list are built-in functions, but I can add them to the pseudocode definitions file. In the case of clamp, the Rust code does not directly use a list, but talking to Mike we settled on that the best Python-Rust blend term for this line would be to use a list, instead of the Rust .collect() function that appears in the code.

SalilVadhan replied:
I do think it would be good to add map and state in the pseudocode defs file that it is representing the Rust function collect. Actually doesn't map by definition return a list?  So what is the list(.) function doing?

SalilVadhan replied:
nvm, I see my questions are answered below- but the fact that Grace and I had the same question reinforces that it would be good to add defs of map and list into the pseudocode_defns file

silviacasac replied:
Yes, they are added now

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+        
+        \item \textup{(Domain-metric compatibility).} The domain \texttt{input\_domain} matches one of the possible domains listed in the definition of \texttt{input\_metric}, and likewise \texttt{output\_domain} matches one of the possible domains listed in the definition of \texttt{output\_metric}.
+        
+        \item \textup{(Stability guarantee).} For every pair of elements $v, w$ in \texttt{input\_domain} and for every pair $(\din, \dout)$, where $\din$ and $\dout$ are of type \texttt{u32}, if $v,w$ are $d_{in}$-close under \texttt{input\_metric} and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$ are $d_{out}$-close under \texttt{output\_metric}.
```
A more general phrasing: "where din is the associated type for input_metric and dout is the associated type for output_metric".  That way you can use the same phrasing for all transformations.

silviacasac replied:
Yes this is better, we wanted a more general formulation -- changed.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+\end{enumerate}
+In all three cases, the returned value of type \T is contained in the interval \texttt{[L, U]}. Hence, the vector $\function(v)$ returned in line~\ref{line:map} of the pseudocode is an element of \texttt{output\_domain}.
+
+Lastly, the necessary condition that \texttt{L} $\leq$ \texttt{U} is checked in line~\ref{line:excep} of the pseudocode, hence correctness is guaranteed if no exception is raised. Both \texttt{L} and \texttt{U} have type \texttt{T} by their precondition requirement. Both the definition of \texttt{IntervalDomain} and that of the \texttt{clamp} function (line~\ref{line:clamp} in the pseudocode, whi ...
```
I'd write "require that T implement TotalOrd, which holds by the preconditions."

SalilVadhan replied:
also this text is referring to IntervalDomain which is currently not used in the pseudocode.  I assume this will change when the defs of the domains move back into the pseudocode.

SalilVadhan replied:
(that is, the pseudocode will then use "IntervalDomain")

silviacasac replied:
Changed, now that the defs of the domains and metrics are back into the pseudocode.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+%\silvia{So ``privacy relation" for measurements $=$ ``stability relation" for transformations, but both are providing the DP guarantee. But maybe ``DP guarantee" is not the best name}
+
+\smallskip
+\textbf{(Domain-metric compatibility).}  For \texttt{MakeClamp}, both cases correspond to showing that \texttt{VectorDomain(T)} is compatible with symmetric distance. This follows directly from the definition of symmetric distance, as stated in \href{https://www.overleaf.com/project/60d214e390b337703d200982}{``List of definitions used in the proofs"}. The \textit{appropriate output domain propert ...
```
I don't think you need to refer to the appropriate output domain property, which involves the function.  Here you just need to refer to the line that defines "output_domain = ..." (currently in preconditions) and check that it is compatible with the line "output_metric = ..."

silviacasac replied:
Ah right, I was thinking about this as showing compatibility between output_metric and the domain of function(v) (which should be the same as output_domain by 1.). But we are showing compatibility with output_domain, which is directly defined in the pseudocode. Changed.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+\smallskip
+\textbf{(Stability guarantee).} Throughout the stability guarantee proof, we can assume that $\function(v)$ and $\function(w)$ are in the correct output domain, by the \textit{appropriate output domain property} shown above. 
+
+Since by assumption $\Relation(\din, \dout) = \True$, by the \texttt{MakeClamp} stability relation (as defined in line~\ref{line:rel} in the pseudocode), we have that $\din \leq \dout$. Moreover, $v, w$ are assumed to be $\din$-close. By the definition of the symmetric difference metric, this is equivalent to stating that $d_{Sym}(v, w) = |\MultiSet(v) \Delta \MultiSet(w)| \leq \din$.
```
should add the definition of the histogram notation to the proof_defns file and refer to that rather than the OpenDP programming framework paper

silviacasac replied:
The definition of histogram notation is already part of the proof_defns file (see Definitions 2.4 and 2.5); I was citing the programming framework for completeness, but I will take it out.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+\textbf{(Domain-metric compatibility).}  For \texttt{MakeClamp}, both cases correspond to showing that \texttt{VectorDomain(T)} is compatible with symmetric distance. This follows directly from the definition of symmetric distance, as stated in \href{https://www.overleaf.com/project/60d214e390b337703d200982}{``List of definitions used in the proofs"}. The \textit{appropriate output domain propert ...
+
+\smallskip
+\textbf{(Stability guarantee).} Throughout the stability guarantee proof, we can assume that $\function(v)$ and $\function(w)$ are in the correct output domain, by the \textit{appropriate output domain property} shown above. 
```
I think the following will give a simpler proof.  Whenever we have a function from Vector Domains to Vector Domains of the form function(data)=list(map(f,data)), then h_{function(v)}(z) = sum_{y in f^{-1}(z)) h_v(y).  And then apply triangle inequality and that the sets f^{-1}(z) form a partition of the domain of f.  This will avoid a need to do a case analysis, and we can use the same proof for all of transformations of this form.  In fact, this can be a general lemma in the proof_defns file that we cite, or a proof for "make_rowtransform" (which I believe that Mike may have written, but which will require a precondition on the function f).

SalilVadhan replied:
the requirement is that f should be a "pure" function, i.e. a function in the standard mathematical sense, not a piece of code that has side effects or can access some global state.

gracetian6 replied:
I am typing up proof of stability relations for row by row transformations. Would it make more sense for this to be in the [list of definitions in proofs](https://github.com/opendp/whitepapers/blob/proof-defns/proof-defns/proof_defns.pdf) or a separate transformation proof for `make_row_by_row`?

silviacasac replied:
Decided on August 4: clamp will not include this simplified row by row proof until the Rust code for clamp actually calls the `make_row_by_row` transformation inside of it. But see e.g., `is_equal` for a use case of the `make_row_by_row` transformation.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,331 @@
+as we wanted to show.
+\end{proof}
+
+\silvia{Maybe add domain of $z$ below the sum?}
```
yes, that would be good!

gracetian6 commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,248 @@
+            return max(min(x, U), L)
+        return list(map(clamp, data)) |\label{line:map}|
+    
+    return Transformation(input_domain, output_domain, function, input_metric, output_metric, stability_relation)
```
in line 148, stability_relation is not defined in pseudo code. Maybe change name of Relation or instantiate stability_relation variable?

silviacasac replied:
Ah, thanks for the catch! I thought I had changed it everywhere

gracetian6 commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,248 @@
+\end{theorem}
+
+\begin{proof}
+\textbf{(Appropriate output domain).} In the case of \texttt{MakeClamp}, this corresponds to showing that for every vector $v$ of elements of type \texttt{T}, $\function(v)$ is a vector of elements of type \texttt{T} which are contained in the interval \texttt{[L, U]}. For that, we need to show two things: first, that \texttt{function(v)} has type \texttt{Vec(T)}.
+Second, that they belong to the interval \texttt{[L, U]}.
+
+Firstly, that $\function(v)$ has type \texttt{Vec(T)} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(T)} and returns an element of type \texttt{Vec(T)}. If the Rust code compiles correctly, then the type c ...
+
+Secondly, we need to show that the vector entries belong to the interval \texttt{[L, U]}. This follows from the definition of \texttt{clamp} in line \ref{line:clamp}. According to line \ref{line:clamp} in the pseudocode, there are 3 possible cases to consider:
```
very well organized; it really helps that you broke down each case to consider with firstly, secondly, etc.

gracetian6 commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,248 @@
+        return d_out >= d_in*1
+    
+    def function(data: Vec(T)) -> Vec(T): |\label{line:fn}|
+        def clamp(x: T) -> T: |\label{line:clamp}|
+            return max(min(x, U), L)
+        return list(map(clamp, data)) |\label{line:map}|
```
Why do we use list()? By definition of map, it should apply to clamp function elementwise to the entire input vector. https://en.wikipedia.org/wiki/Map_(higher-order_function)#:~:text=In%20many%20programming%20languages%2C%20map,when%20considered%20in%20functional%20form.

gracetian6 replied:
Nvm, Silvia explained that the reaosn is because

Mike: “2. returning a map has a slightly different meaning. list(map(...)) or [clamp(L, U, v) for v in arg] is more accurate. In rust, the .collect exhausts the iterator into a vector. So the analogue here is list or the list comprehension (not a generator comprehension). It would be cool if we chained iterators, but that makes other things more complicated (edited)”

silviacasac replied:
Right so we want to return a vector (for appropriate output domain) hence we use the Python `list`, not a map

Shoeboxam commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,252 @@
+            return max(min(x, U), L)
+        return list(map(clamp, data)) |\label{line:map}|
+    
+    return Transformation(input_domain, output_domain, function, input_metric, output_metric, stability_relation = Relation)
```
Why the special naming "Relation" instead of "stability_relation"? The code makes sense. Are you guys planning on switching type hints to the [type hint syntax](https://www.python.org/dev/peps/pep-0484/)? I prefer the typing syntax for hints as it looks less like a function invocation. It's a pretty minor pick, basically `Vec(T)` -> `Vec[T]`. Also, are we keeping `Vec` from rust or using python's `list` in the pseudocode? The iterator collection uses list, but the type is a `Vec`. I would change one or the other to be consistent.

silviacasac replied:
Regarding the special naming of "Relation": I mainly did this to save space in Section 2. Proof when proving the stability relation, so that only "Relation" appears in the text instaed of "stability_relation". And also so that the theorem statement for the stability/privacy guarantee looks the same for measurements, where we would have `privacy_relation = Relation`. But if you think this is not enough to change it I can add `stability_relation` back, no problem.

silviacasac replied:
About the type hint syntax, are you saying that the correct function type signing for clamp is `def function(data: Vec(T)) -> Vec[T]`? How so? So initially we had `Vec(T) -> Vec(T)` but you said that this was not correct, and so we introduced the rule "Parentheses, `()`, are used to create an instance of a domain. Square brackets, `[]`, are used to describe the type of a domain. For example, `AllDomain(i8)` is the domain of all values of type `i8.` However, the type of `AllDomain(i8)` -- the domain itself, not the elements of the domain -- is `AllDomain[i8];` note the square brackets. This typing style is inspired by the notation used in Python; see \url{https://docs.python.org/3/library/typing.html}." (this is now included in the pseudocode definitions).

silviacasac replied:
Regarding Vec vs list: so I thought we wanted the pseudocode to look as Python as possible, which is why I wrote it in list format. But then it seemed not correct to change the input and output domains, because the `VectorDomain` formulation is important and appears everywhere, and also for showing correct output domain. When you say you would change either or the other are you proposing to either: 1. Change line 148 to `return Vec(map(clamp, data))` and keep the domains as they are, or 2. To keep line 148 as it is and change the domains to something different? Approach 2 seems worse I think, because that is inventing new domains that don't exist in the Rust library. And could we not just say that `list` has type `Vec` in the list of definitions to resolve this? But perhaps approach 1 is better then, although less Python

Shoeboxam replied:
1. I think the part throwing me is that camel-case code in python usually represents classes. I have a soft preference towards `relation` or `stability_relation`.
2. No, the arrow in my statement was meant to communicate "replace Vec(T) with Vec[T]". The current syntax in the pseudocode is `def function(data: Vec(T)) -> Vec(T):`. The revised syntax would be `def function(data: Vec[T]) -> Vec[T]:`. The type hints should use the type syntax, because you're not making instances of a Vec within the signature.

3. You're right, there's no win here. I think we leave Vec/list inconsistent unless someone else has strong feelings.

silviacasac replied:
1. That makes sense, I had not considered the implications of having it capitzalied. I will change all of them to `relation` then.
2. Ah okay, I thought the TeX had already been changed to `Vec(T)`, but it is fixed now.
3. Okay!

Shoeboxam commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,252 @@
+    input_metric = SymmetricDistance()
+    output_metric = SymmetricDistance()
+    
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
```
Will switching to IntDistance cause headaches elsewhere? I'm thinking no, but sometimes making things more flexible complicates things. IntDistance is just a type alias, the user can't change it, but having it behind an alias makes it easy to change across the library. The naming could be adjusted, as always. It might make sense to add an explicit list of types IntDistance can be- basically all primitive int types.

silviacasac replied:
Oh right you had changed the associated type of symmetric distance to `IntDistance`, right? But then set `IntDistance = u32` -- so you say to add an explicit list of types `IntDistance` can be, but isn't it only used for `u32` in symmetric distance no? And are you saying that I should change the above to `def Relation(d_in: IntDistance, d_out: IntDistance)`? Or perhaps not yet because it does not appear in the clamp Rust code?

silviacasac replied:
Okay I think you are saying to write `def Relation(d_in: IntDistance, d_out: IntDistance)` instead, but then should `IntDistance = u32` be specified anywhere in the clamp pseudocode or do we add the explicit list of types for IntDistance in the definitions for the pseudocode document?

Shoeboxam replied:
It would be nice to pull this u32 constant out in favor of an alias, but it does seem to cause headaches--
1. The name is much longer
1. We use all sorts of traits/methods on IntDistance that aren't qualified in the trait bounds because it's a type alias, not a generic. I wouldn't want to have to pull these details down into proofs everywhere.

I vote we don't bother making changes at this point. If we did, I would advocate:
1. a better alias name 
2. the alias definition only makes a definitive statement that it is defined to be u32 
3. include a weaker statement that any type from a list of int types should be isomorphic, but not give explicit guarantees.

Maybe a better approach is to just add the statement in (3) to the definition of SymmetricDistance, and leave the proofs as-is.

silviacasac replied:
Okay, for now I will leave `u32` then.

silviacasac commented:
Remaining things to double-check (all minor):
- `Vec` vs `list`
- `IntDistance` vs `u32`
- `Vec(T)` vs `Vec[T]`

Will require future edits if `make_row_by_row` is added to the clamp Rust code in the library.

Edit: the remaining things have been double-checked.

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,260 @@
+\section{Algorithm Implementation}
+
+\subsection{Code in Rust}
+The current OpenDP library contains the \texttt{make\_clamp\_vec} function implementing the clamping function. This is defined in lines 25-38 of the file \texttt{manipulation.rs} in the Git repository\footnote{As of June 16, 2021.}  (\url{https://github.com/opendp/opendp/blob/58feb788ec78ce739caaf3cad8471c79fd5e7132/rust/opendp/src/trans/manipulation.rs#L25-L38}).
```
comment on the Rust code (should probably be a comment on manipulation.rs): 1-stability and Lipschitz constant of 1 are the same thing (so "as well as" should be "i.e.")

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,260 @@
+
+Since then, the code has been updated to include a more general clampable domain to abstract together parts of the vector and scalar cases:\footnote{As of July 28, 2021.} \url{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/clamp.rs#L28-L44} in conjunction with \url{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/clamp.rs#L74-L85}. We decided not to refer to the  ...
+
+\begin{figure}[ht]
+    \includegraphics[width=16cm]{clamp3.png}
+    \centering
+    \label{fig:code}
+\end{figure}
+
+\begin{figure}[ht]
+    \includegraphics[width=16cm]{clamp4.png}
+    \centering
+    \label{fig:code}
+\end{figure}
```
why include this code if the proof is only going to be for make_clamp_vec?

silviacasac replied:
I took it out

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,260 @@
+        
+        \item \textup{(Domain-metric compatibility).} The domain \texttt{input\_domain} matches one of the possible domains listed in the definition of \texttt{input\_metric}, and likewise \texttt{output\_domain} matches one of the possible domains listed in the definition of \texttt{output\_metric}.
+        
+        \item \textup{(Stability guarantee).} For every pair of elements $v, w$ in \texttt{input\_domain} and for every pair $(\din, \dout)$, where $\din$ is of the associated type for \texttt{input\_metric} and $\dout$ is the associated type for \texttt{output\_metric}, if $v,w$ are $\din$-close under \texttt{input\_metric} and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$  ...
```
minor nit "dout is the associated type" -> "dout is of the associated type" (actually "dout has the associated type" reads more nicely)

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,260 @@
+\subsection{Symmetric Distance}
+\begin{theorem}
+    For every setting of the input parameters \texttt{(L, U)} to \texttt{MakeClamp} such that the given preconditions
+    hold, the transformation returned by \texttt{MakeClamp} has the following properties:
```
shouldn't we say "MakeClamp raises a exception (at compile time or run time) or returns a valid transformation."?  (And then we can/should put the def of what constitutes a valid transformation in either the pseudocode_defns or proof_defns file (not sure which).)

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,260 @@
+\end{enumerate}
+In all three cases, the returned value of type \T is contained in the interval \texttt{[L, U]}. Hence, the vector $\function(v)$ returned in line~\ref{line:map} of the pseudocode is an element of \texttt{output\_domain}.
+
+Lastly, the necessary condition that \texttt{L} $\leq$ \texttt{U} is checked when declaring \texttt{output\_domain = VectorDomain(IntervalDomain(L, U))} in line~\ref{line:output} of the pseudocode. This check already exists via the construction of \texttt{IntervalDomain}, which returns an error if \texttt{L} $>$ \texttt{U}. Both \texttt{L} and \texttt{U} have type \texttt{T} by the type signature ...
```
I think the fact that IntervalDomain raises an exception when L>U should be stated at the beginning of the proof, before item 1.  Otherwise much of what follows doesn't make sense (e.g. case analysis above for clamp).

SalilVadhan commented on clamp/clamp.tex:
```diff
@@ -0,0 +1,260 @@
+as we wanted to show.
+\end{proof}
+
+\silvia{Flag: will change to the simplified general proof scheme for row-by-row transformations that Prof. Vadhan suggested in the Git repo once the \texttt{make\_row\_by\_row} transformation is included as part of the clamp transformation code.}
```
I think it's worth doing that here in this proof even prior to make_row_by_row.  It is much simpler and shorter than what's above.
